source-academy/frontend
# Mirror editor updates to the file system & save playground editor tab state to localstorage

### Description

Sorry, this PR is a bit unfocused because integration is rather messy and I'm figuring things out as I go along.

Added the `filePath` field to the `EditorTabState:
```ts
export type EditorTabState = {
  readonly filePath?: string;
  readonly value: string;
  readonly highlightedLines: HighlightedLines[];
  readonly breakpoints: string[];
  readonly newCursorPosition?: Position;
};
```

If an editor tab is associated with a file in the file system, all updates to it are written to the file. This is done by adding a saga that fires whenever the `UPDATE_EDITOR_VALUE` action is dispatched.

Also saved all of the playground editor tab states to localstorage so that the open tabs are preserved across visits to the Source Academy.

Finally, added event handlers when double clicking or right clicking and selecting 'Open' on files in the file system view. Currently, only a message is logged to the console because this is waiting on another PR that enables the `EditorContainer` to display multiple editors.

Part of #2176.

### Type of change


### How to test

Revert the changes made in https://github.com/source-academy/frontend/commit/35e7bc281458a215bd653a8ca6de590ef51a9ce8 to access the file system view. Currently, there should be no visible changes in behaviour, although every update to an editor tab that is associated with a file will trigger a corresponding update to the underlying file system. Additionally, double clicking on a file or right-clicking to open a file will result in a message being output to the console.
